[Serve] Support http_location=FixedNumber

simon-mo opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?
This PR adds a new undocumented options for setting a fixed number for HTTP proxies. Serve will deterministically sample from the set of available nodes to place http proxy on. User API:
```python
serve.start("http_options": {"location": "FixedNumber", "fixed_number_replicas": 3})
```

<!-- Please give a short summary of the change and the problem this solves. -->

## Related issue number

<!-- For example: "Closes #1234" -->

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

edoakes commented on python/ray/serve/config.py:
```diff
@@ -214,6 +214,7 @@ class DeploymentMode(str, Enum):
     NoServer = "NoServer"
     HeadOnly = "HeadOnly"
     EveryNode = "EveryNode"
+    FixedNumber = "FixedNumber"
```
Hmm I'm pretty sure that DeploymentMode shows up in the API docs -- is there any way to hide this option?

simon-mo replied:
![image](https://user-images.githubusercontent.com/21118851/134081399-f6dd3237-88d0-469a-9602-3088c3239318.png)
they are not in fact. Because we hand wrote the doc in `serve.start` doc string and I purposely didn't document it there.

edoakes commented on python/ray/serve/http_state.py:
```diff
@@ -54,9 +55,24 @@ def _get_target_nodes(self) -> List[Tuple[str, str]]:
+                    "HTTP proxy per node.")
+                num_replicas = len(target_nodes)
+
+            # Seed the random state so sample is deterministic.
+            # i.e. it will always return the same set of nodes.
+            random.seed(0)
```
Can you please add a unit test for this behavior? e.g., pull the node-selecting logic out into a utility and add some tests for it.

simon-mo replied:
Done.

edoakes reviewed (commented):
Looks good. Please add unit tests for the node selection logic.

simon-mo commented:
@edoakes I added the unit test. I also added a parameter make the seed for the random selection configurable. This is because on a 100 node cluster, with multiple Serve deployments, we want don't want every set of http proxy to land on the same set of nodes.
